Modules/StudyProgramme/classes/class.ilStudyProgrammeMembersTableGUI.php

```diff
@@ -121,40 +98,92 @@ protected function getLinkTargetForAction($a_action, $a_prgrs_id, $a_ass_id) {
-							   ."  LEFT JOIN object_data cmpl_obj ON cmpl_obj.obj_id = prgrs.completion_by"
-							   ." WHERE prgrs.prg_id = ".$this->db->quote($a_prg_id, "integer")
-							   );
+		$query = "SELECT prgrs.id prgrs_id"
+				   ."     , pcp.firstname"
+				   ."     , pcp.lastname"
+				   ."     , pcp.login"
+				   ."     , prgrs.points"
+				   ."     , if(prgrs.status = ".ilStudyProgrammeProgress::STATUS_ACCREDITED.", prgrs.points, prgrs.points_cur) points_current"
+				   ."     , prgrs.last_change_by"
+				   ."     , prgrs.status"
+				   ."     , blngs.title belongs_to"
+				   ."     , cmpl_usr.login accredited_by"
+				   ."     , cmpl_obj.title completion_by"
+				   ."     , prgrs.assignment_id assignment_id"
+				   ."     , ass.root_prg_id root_prg_id"
+				   ."  FROM ".ilStudyProgrammeProgress::returnDbTableName()." prgrs"
+				   ."  JOIN usr_data pcp ON pcp.usr_id = prgrs.usr_id"
+				   ."  JOIN ".ilStudyProgrammeAssignment::returnDbTableName()." ass"
+				   			 ." ON ass.id = prgrs.assignment_id"
+				   ."  JOIN object_data blngs ON blngs.obj_id = ass.root_prg_id"
+				   ."  LEFT JOIN usr_data cmpl_usr ON cmpl_usr.usr_id = prgrs.completion_by"
+				   ."  LEFT JOIN object_data cmpl_obj ON cmpl_obj.obj_id = prgrs.completion_by"
+				   ." WHERE prgrs.prg_id = ".$this->db->quote($a_prg_id, "integer");
+
+		if($order_coloumn !== null) {
+			$query .= " ORDER BY $order_coloumn";
+
+			if($order_direction !== null) {
+				$query .= " $order_direction";
+			}
+		}
+
+
+		if($limit !== null) {
+			$query .= " LIMIT ".$this->db->quote($limit, "integer");
+		}
+
+		if($offset !== null) {
+			$query .= " OFFSET ".$this->db->quote($offset, "integer");
+		}
+		$res = $this->db->query($query);
 	
 		$members_list = array();
 		while($rec = $this->db->fetchAssoc($res)) {
 			$rec["actions"] = ilStudyProgrammeUserProgress::getPossibleActions(
 										$a_prg_id, $rec["root_prg_id"], $rec["status"]);
+
+			if ($rec["status"] == ilStudyProgrammeProgress::STATUS_COMPLETED) {
+				// If the status completed and there is a non-null completion_by field
+				// in the set, this means the completion was achieved by some leaf in
+				// the program tree.
+				if (!$rec["completion_by"]) {
+					require_once("Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgress.php");
+					$prgrs = ilStudyProgrammeUserProgress::getInstanceForAssignment( $this->prg_obj_id
+																					  , $rec["assignment_id"]);
+					$rec["completion_by"] = implode(", ", $prgrs->getNamesOfCompletedOrAccreditedChildren());
+				}
+			}
+			else if($rec["status"] == ilStudyProgrammeProgress::STATUS_ACCREDITED) {
+				$rec["completion_by"] = $rec["accredited_by"];
+			}
+
 			$members_list[] = $rec;
 		}
 		return $members_list;
 	}
+
+	protected function countFetchData($a_prg_id) {
+		// TODO: Reimplement this in terms of ActiveRecord when innerjoin
+		// supports the required rename functionality
+		$query = "SELECT count(prgrs.id) as cnt"
```

Comment:
Ich glaube es wäre gut, wenn wir die Queries aus fetchData und countFetchData irgendwie zusammen erstellen können, zumindestens die Teile die sich gleichen. Sonst müssen wir in Zukunft immer an zwei Stellen ändern.